model_compression_toolkit/graph_builder/common/base_graph_builder.py

```diff
@@ -0,0 +1,112 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+from abc import ABC, abstractmethod
+from typing import Any, Callable, Optional
+
+from model_compression_toolkit.core.common.visualization.tensorboard_writer import TensorboardWriter
+
+from model_compression_toolkit.target_platform_capabilities.targetplatform2framework.framework_quantization_capabilities import \
+    FrameworkQuantizationCapabilities
+
+from model_compression_toolkit.core.common import Graph
+
+
+class BaseGraphBuilder(ABC):
+    """
+    Abstract base class for converting framework-specific models into a framework-agnostic computational graph,
+    and applying standard graph transformations.
+
+    Subclasses must implement methods to:
+    1. Convert a model to a graph representation.
+    2. Apply optional graph simplifications and preprocessing.
+    """
+
+    def build_graph(self,
+                    model: Any,
+                    representative_dataset: Callable = None,
+                    fqc: FrameworkQuantizationCapabilities = None,
+                    tensorboard_writer: TensorboardWriter = None,
+                    linear_collapsing: bool = True,
+                    residual_collapsing: bool = True,
+                    relu_bound_to_power_of_2: bool = False):
+        """
+        Converts a model to a graph and applies selected graph transformations. Optionally logs the graph
+        before and after transformation using TensorBoard.
+
+        Args:
+            model (Any): The framework-specific model to convert.
+            representative_dataset (Callable, optional): A callable that yields representative inputs for tracing
+                (used in some frameworks like PyTorch).
+            fqc (FrameworkQuantizationCapabilities, optional): Object containing quantization capabilities of the framework.
+                If provided, it's stored in the graph.
+            tensorboard_writer (TensorboardWriter, optional): If provided, logs the initial and transformed graphs.
+            linear_collapsing (bool): Whether to collapse sequences of linear layers (e.g., dense or conv).
+            residual_collapsing (bool): Whether to collapse residual connections.
+            relu_bound_to_power_of_2 (bool): If True, replaces ReLU activation bounds with nearest power-of-2 (for HW-friendliness).
+
+        Returns:
+            Graph: The transformed, simplified computational graph.
+        """
+        graph = self.convert_model_to_graph(model, representative_dataset)
+        if tensorboard_writer is not None:
+            tensorboard_writer.add_graph(graph, 'initial_graph')
+
+        if fqc:
```

Comment:
I don't think that fqc should be an argument to this module, since it is quantization-related.
If we must still store it in the graph, then it supposed to be the "quantization_preparation" responsibility